src/engine/PiecewiseLinearConstraint.h

```diff
@@ -31,6 +31,17 @@ class ITableau;
 class InputQuery;
 class String;
 
+enum PhaseStatus : unsigned {
+    PHASE_NOT_FIXED = 0,
+    RELU_PHASE_ACTIVE = 1,
+    RELU_PHASE_INACTIVE = 2,
+    ABS_PHASE_POSITIVE = 3,
+    ABS_PHASE_NEGATIVE = 4,
+    SIGN_PHASE_POSITIVE = 5,
+    SIGN_PHASE_NEGATIVE = 6
+
```

Comment:
Remove empty line

Reply:
All PiecewiseLinearConstraints have a phase/case, either a fixed number of them (ReLU, Sign, Abs) or a flexible number of them (Max, Disjunction). The first  group cares about named elements of the enumeration, while the second group cares about their numerical value, and a fixed size enumeration caters to both. Regardless of the subclass, the role of these fields is the same and becomes further relevant in context-dependent implementation for detection of implications, contradictions, etc.   

The idea is to generalize this field to avoid repetition of identical code in child classes, and level the implementation so that a single intermediate child class implement the context-dependent behavior, while retaining existing functionality for preprocessing purposes. (The intermediate class can be merged with the parent class at a later date). 

Implementation-wise, we can make PiecewiseLinearConstraint a template class and provide the individual PhaseStatus enumerations as the parameter. However, instead of one global enumeration, there would be several smaller enumerations, which would be outside of their child template classes. This solution has some small upsides for type-checking, but adds some overhead for implementation. 

For general overview, the next PR adds ContextDependentPiecewiseLinearConstraint, which inherits PiecewiseLinearConstraint and  introduces context-dependent and lazily backtracking members _cdConstraintActive, _cdPhaseStatus and uniformly implement relevant methods which use parent class members until the context-based members are initialized. Current architecture would require that child classes implement individually context-dependent behavior, which would be undesirable given the delicacy of context-dependent data structures and causing additional code repetition. Finally, this approach allows for simple transition to context-based architecture (when the time comes), by inheriting the ContextDependentPiecewiseLinearConstraint class and transitively the PiecewiseLinearConstraint class. If needed I can include that PR with this.